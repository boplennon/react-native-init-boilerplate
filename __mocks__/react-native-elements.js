
module.exports = ({
  FormValidationMessage: 'View',
  SocialIcon: 'Button',
  FormLabel: 'Text',
  Icon: 'View'
});

