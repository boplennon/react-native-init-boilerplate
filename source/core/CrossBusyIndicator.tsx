import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import {
  BallIndicator,
  BarIndicator,
  DotIndicator,
  MaterialIndicator,
  PacmanIndicator,
  PulseIndicator,
  SkypeIndicator,
  UIActivityIndicator,
  WaveIndicator,
} from 'react-native-indicators';

import styles, { Colors } from '../styles';

/**
 * Type of indicator
 *
 * @export
 * @enum {string}
 */
export enum IndicatorType {
  BallIndicator = 'BallIndicator',
  BarIndicator = 'BarIndicator',
  DotIndicator = 'DotIndicator',
  MaterialIndicator = 'MaterialIndicator',
  PacmanIndicator = 'PacmanIndicator',
  PulseIndicator = 'PulseIndicator',
  SkypeIndicator = 'SkypeIndicator',
  UIActivityIndicator = 'UIActivityIndicator',
  WaveIndicator = 'WaveIndicator',
}

/**
 * Generates a spinner of the provided type
 * @param param0 type and style
 */
export const Spinner = ({
  type,
  style = null,
  color = Colors.CrossLightBlue,
}: {
  type: IndicatorType;
  style?: any;
  color?: string;
}) => {
  switch (type) {
    case 'BallIndicator':
      return (
        <BallIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'BarIndicator':
      return (
        <BarIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'DotIndicator':
      return (
        <DotIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'PacmanIndicator':
      return (
        <PacmanIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'PulseIndicator':
      return (
        <PulseIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'SkypeIndicator':
      return (
        <SkypeIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'UIActivityIndicator':
      return (
        <UIActivityIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    case 'WaveIndicator':
      return (
        <WaveIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
    default:
      return (
        <MaterialIndicator
          // @ts-ignore
          style={style}
          color={color}
        />
      );
  }
};

/**
 * Describes the props BusyIndicator accepts
 *
 * @export
 * @interface IBusyIndicatorProps
 */
export interface IBusyIndicatorProps {
  /**
   * Determines if the indicator is animating or not. Default value is true
   *
   * @type {boolean}
   * @memberof IBusyIndicatorProps
   */
  isBusy: boolean;
  /**
   * Optional message to show while busy. Defaults to empty string.
   *
   * @type {string}
   * @memberof IBusyIndicatorProps
   */
  message?: string;
  /**
   * Type of spinner to use
   *
   * @type {IndicatorType}
   * @memberof IBusyIndicatorProps
   */
  type?: IndicatorType;
  /**
   * Test id to assign this component
   *
   * @type {string}
   * @memberof IBusyIndicatorProps
   */
  testID?: string;
  /**
   * User pressed cancel
   */
  onCancel?: () => void;
}

/**
 * Displays a busy indicator (spinner) when the parameter is true
 * @param {bool} isBusy determines if the indicator is animating or not. Default value is true
 * @param {string} message optional message to show while busy. Defaults to empty string.
 * @param {IndicatorType} type of spinner.Defaults to "MaterialIndicator"
 */
const CrossBusyIndicator = ({
  isBusy = true,
  message = '',
  type = IndicatorType.MaterialIndicator,
  testID = '1',
  onCancel,
}: IBusyIndicatorProps) =>
  isBusy ? (
    <View style={styles.overlay} testID={testID}>
      <View style={[styles.absoluteCentered]}>
        <View style={[styles.columnContentTopCenter]}>
          <Spinner type={type} style={[styles.spinner]} />
        </View>
        <View style={styles.columnContentTopCenter}>
          <Text style={styles.textSpinner}>{message}</Text>
          {onCancel ? (
            <TouchableOpacity onPress={onCancel}>
              <Text style={[styles.textSpinner, { color: Colors.CancelButton }]}>
                Avbryt
              </Text>
            </TouchableOpacity>
          ) : null}
        </View>
      </View>
    </View>
  ) : null;

export default CrossBusyIndicator;
