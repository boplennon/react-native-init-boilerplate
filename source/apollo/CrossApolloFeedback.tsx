import * as React from 'react';
import { View } from 'react-native';
import { FormValidationMessage } from 'react-native-elements';
import { styles } from '../styles';
import CrossBusyIndicator from '../core/CrossBusyIndicator';

interface IProps {
  error?: Error;
  loading?: boolean;
  loadingMessage?: string;
}

/**
 * Takes loading and error from GraphQL Apollo queries / mutations and displays the appropriate feedback
 * @param param0 {@link IProps} props
 */
export const CrossApolloFeedback = ({
  loading,
  error = undefined,
  loadingMessage = undefined,
}: IProps) => {
  if (loading === true) {
    console.log('** MOBILE: Loading ** ' + loadingMessage);
  }

  return (
    <View style={styles.container}>
      {error ? (
        <FormValidationMessage>{error.message}</FormValidationMessage>
      ) : null}
      {loading === true ? (
        <CrossBusyIndicator isBusy message={loadingMessage} />
      ) : null}
    </View>
  );
};

export default CrossApolloFeedback;
