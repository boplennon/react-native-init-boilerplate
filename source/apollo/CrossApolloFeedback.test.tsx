import React from 'react';
import TestRenderer from 'react-test-renderer';
import CrossApolloFeedback from './CrossApolloFeedback';
import { FormValidationMessage } from 'react-native-elements';
import CrossBusyIndicator from '../core/CrossBusyIndicator';

jest.unmock('react-native');
jest.unmock('react-native-elements');
jest.mock('../core/CrossBusyIndicator', () => 'View');

const wrapper = (busy: boolean, error: Error) =>
  TestRenderer.create(
    <CrossApolloFeedback
      loading={busy}
      error={error}
      loadingMessage="testing"
    />
  );

describe('components', () => {
  describe('<CrossApolloFeedback />', () => {
    it('should render self and subcomponents', () => {
      const json = wrapper(false, null).toJSON();
      expect(json).toMatchSnapshot();
    });

    it('given loading is true should have <CrossBusyIndicator />', () => {      
      const comp = wrapper(true, undefined);
        const childComponent = comp.root.findByType(CrossBusyIndicator);
        expect(childComponent).toBeDefined();
    });

    it('given error is passed should have <FormValidationMessage />', () => {      
      const comp = wrapper(false, new Error('testing'));
        const childComponent = comp.root.findByType(FormValidationMessage);
        expect(childComponent).toBeDefined();
    });
  });
});
