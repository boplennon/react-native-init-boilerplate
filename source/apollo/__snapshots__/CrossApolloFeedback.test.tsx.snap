// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CrossApolloFeedback should render self and subcomponents 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingTop": 20,
    }
  }
/>
`;
